prov/rxm/src/rxm_ep.c

```diff
@@ -925,10 +936,63 @@ rxm_ep_lmt_tx_send(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
 }
 
 static inline size_t
-rxm_ep_sar_estimate_segments_num(struct rxm_ep *rxm_ep, size_t data_len)
+rxm_ep_sar_calc_seg_len(struct rxm_ep *rxm_ep, size_t index, size_t remain_len)
+{
+	size_t seg_len;
+
+	seg_len = (index <= rxm_ep->sar.max_calc_seg_no) ?
+		  RXM_SAR_DIVIDER << index :
+		  rxm_ep->rxm_info->tx_attr->inject_size;
+	if (seg_len >= remain_len)
+		seg_len = remain_len;
+
+	return seg_len;
+}
+
+static inline size_t
+rxm_ep_sar_calc_segs_cnt_limit(struct rxm_ep *rxm_ep, size_t size_len)
+{
+	size_t segs_cnt_lim = 0;
+
+	while (size_len)
+		size_len -= rxm_ep_sar_calc_seg_len(rxm_ep, segs_cnt_lim++,
+						    size_len);
+
+	return segs_cnt_lim;
+}
+
+static inline size_t
+rxm_ep_sar_calc_size_limit(struct rxm_ep *rxm_ep, size_t max_segs_cnt)
+{
+	size_t size_limit = 0, segs_cnt = 0;
+
+	while (segs_cnt < max_segs_cnt) {
+		size_limit += (segs_cnt <= rxm_ep->sar.max_calc_seg_no) ?
+			      RXM_SAR_DIVIDER << segs_cnt :
+			      rxm_ep->rxm_info->tx_attr->inject_size;
+		segs_cnt++;
+	}
+
+	return size_limit;
+}
+
+static inline size_t
+rxm_ep_sar_calc_segs(struct rxm_ep *rxm_ep, size_t data_len,
+		     struct rxm_sar_seg *sar_segs)
 {
-	/* This is rough estimation of the number of the segments to be sent */
-	return (data_len / rxm_ep->rxm_info->tx_attr->inject_size);
+	size_t segs_cnt = 0;
+
+	while (data_len && (segs_cnt < rxm_ep->sar.segs_cnt_limit)) {
+		data_len -= sar_segs[segs_cnt].len =
+			rxm_ep_sar_calc_seg_len(rxm_ep, segs_cnt, data_len);
```

Comment:
This is a *very* odd line of code.  Please split into 2 lines.

Reply:
Done